File: azure_function_ppt_v2/function_app.js
/**
 * Azure Function PowerPoint Generation v2 - Main Entry Point
 * Hybrid architecture with Node.js + PptxGenJS
 */
const { app } = require('@azure/functions');
const fs = require('fs');
const path = require('path');
const { generateSessionId } = require('./config');
const orchestrator = require('./orchestrator');

// Ensure local output directory exists
const LOCAL_OUTPUT_DIR = path.join(__dirname, 'local_output');
if (!fs.existsSync(LOCAL_OUTPUT_DIR)) {
    fs.mkdirSync(LOCAL_OUTPUT_DIR, { recursive: true });
}

app.http('powerpoint_generation_v2', {
    methods: ['POST'],
    authLevel: 'anonymous',
    handler: async (request, context) => {
        context.log('PowerPoint Generation v2 - Processing request');
        
        try {
            // Parse request body
            const requestBody = await request.json();
            
            if (!requestBody) {
                return {
                    status: 400,
                    jsonBody: {
                        status: 'error',
                        error_message: 'Invalid JSON in request body'
                    }
                };
            }

            // Generate session ID if not provided
            const sessionId = requestBody.session_id || generateSessionId();
            context.log(`Processing request with session ID: ${sessionId}`);

            // Process through hybrid orchestrator
            const result = await orchestrator.processRequest(requestBody, sessionId, context);

            // Save PowerPoint file locally if generated
            if (result.response_data?.powerpoint_output?.ppt_data) {
                try {
                    const pptData = result.response_data.powerpoint_output.ppt_data;
                    const filename = result.response_data.powerpoint_output.filename;
                    
                    // Decode base64 and save file
                    const pptBuffer = Buffer.from(pptData, 'base64');
                    const outputPath = path.join(LOCAL_OUTPUT_DIR, filename);
                    
                    fs.writeFileSync(outputPath, pptBuffer);
                    context.log(`PowerPoint file saved to: ${outputPath}`);
                    
                    // Add local save path to response
                    result.response_data.local_save_path = outputPath;
                    
                } catch (saveError) {
                    context.log(`Failed to save PowerPoint file: ${saveError.message}`);
                    result.response_data.local_save_error = saveError.message;
                }
            }

            return {
                status: 200,
                jsonBody: result
            };

        } catch (error) {
            context.log(`Unhandled error: ${error.message}`);
            context.log(`Stack trace: ${error.stack}`);
            
            const sessionId = generateSessionId();
            return {
                status: 500,
                jsonBody: {
                    response_data: {
                        status: 'error',
                        session_id: sessionId,
                        error_message: `Server error: ${error.message}`,
                        service_version: 'v2.0.0',
                        architecture: 'hybrid'
                    }
                }
            };
        }
    }
});

// Health check endpoint
app.http('health_v2', {
    methods: ['GET'],
    authLevel: 'anonymous',
    handler: async (request, context) => {
        return {
            status: 200,
            jsonBody: {
                status: 'healthy',
                service: 'PowerPoint Generation v2',
                version: '2.0.0',
                architecture: 'hybrid',
                timestamp: new Date().toISOString(),
                features: [
                    'PptxGenJS integration',
                    'Hybrid Python/Node.js architecture',
                    'Corporate slide masters',
                    'Advanced table and chart generation',
                    'Faster processing pipeline'
                ]
            }
        };
    }
});

context.log('Azure Function PowerPoint Generation v2 initialized successfully');

File: azure_function_ppt_v2/function_app.py
import azure.functions as func
import logging
import json
import asyncio
import os
import base64
from ppt_orchestrator_v2 import PPTOrchestratorV2

# Local output directory for saving generated presentations
LOCAL_OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "local_output")
if not os.path.exists(LOCAL_OUTPUT_DIR):
    os.makedirs(LOCAL_OUTPUT_DIR)

# Define the function app
app = func.FunctionApp(http_auth_level=func.AuthLevel.ANONYMOUS)

@app.route(route="powerpoint_generation_v2")
async def powerpoint_generation_v2(req: func.HttpRequest) -> func.HttpResponse:
    """
    Main HTTP Trigger for the PowerPoint Generation Service V2.
    Uses Pandoc + Markdown approach for better consistency.
    """
    logging.info('PowerPoint V2 HTTP trigger function processed a request.')

    try:
        # Get the request body as JSON
        req_body = req.get_json()
    except ValueError:
        return func.HttpResponse(
             "Invalid JSON in request body.",
             status_code=400
        )

    # Instantiate the V2 orchestrator
    orchestrator = PPTOrchestratorV2()
    
    # Process the request using the V2 orchestrator
    try:
        response_data = await orchestrator.process_conversation_request(req_body)
    except Exception as e:
        logging.error(f"V2 Orchestrator failed with an unhandled exception: {e}")
        error_response = {
            "response_data": {
                "status": "error",
                "session_id": req_body.get('session_id', 'N/A'),
                "error_message": f"V2 system encountered an unexpected error: {str(e)}",
                "system_version": "V2_Pandoc_Markdown"
            }
        }
        return func.HttpResponse(json.dumps(error_response), status_code=500, mimetype="application/json")

    # Check if the response contains PowerPoint data
    if response_data.get("response_data", {}).get("powerpoint_output"):
        ppt_output = response_data["response_data"]["powerpoint_output"]
        ppt_base64 = ppt_output.get("ppt_data")
        filename = ppt_output.get("filename", "presentation_v2.pptx")
        
        if ppt_base64:
            try:
                # Decode the base64 string back into bytes
                ppt_bytes = base64.b64decode(ppt_base64)
                
                # Create the full path to save the file
                output_filepath = os.path.join(LOCAL_OUTPUT_DIR, filename)
                
                # Write the bytes to a new .pptx file
                with open(output_filepath, "wb") as f:
                    f.write(ppt_bytes)
                
                logging.info(f"SUCCESS: V2 PowerPoint file saved locally to: {output_filepath}")
                
                # Add V2-specific info to the response
                response_data["response_data"]["local_save_path"] = output_filepath
                response_data["response_data"]["file_size_bytes"] = len(ppt_bytes)
                response_data["response_data"]["generation_method"] = "pandoc_markdown"

            except Exception as e:
                logging.error(f"Failed to save V2 file locally: {e}")
                # Add an error note to the response
                response_data["response_data"]["local_save_error"] = str(e)

    # Add V2 system identifier to all responses
    if "response_data" in response_data:
        response_data["response_data"]["system_version"] = "V2_Pandoc_Markdown"

    # Return the complete JSON response to the client
    return func.HttpResponse(
        json.dumps(response_data),
        status_code=200,
        mimetype="application/json"
    )